ElectricCar.cpp

```diff
@@ -1,16 +1,48 @@
 #include "ElectricCar.hpp"
 #include <iostream>
 
-ElectricCar::ElectricCar(ElectricEngine* engine)
-    : engine_(engine)
-{
-    std::cout << __FUNCTION__ << std::endl;
+ElectricCar::ElectricCar(std::unique_ptr<IEngine> engine)
+    : engine_(std::move(engine)) {
+  std::cout << __FUNCTION__ << ": " << __FILE__ << ": " << __LINE__
+            << std::endl;
 }
 
-ElectricCar::~ElectricCar()       { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::turnLeft()      { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::turnRight()     { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::brake()         { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::accelerate(int) { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::charge()        { std::cout << __FUNCTION__ << std::endl; }
+ElectricCar::~ElectricCar() {
+  std::cout << __FUNCTION__ << ": " << __FILE__ << ": " << __LINE__
+            << std::endl;
+}
+void ElectricCar::turnLeft() const {
+  std::cout << __FUNCTION__ << ": " << __FILE__ << ": " << __LINE__
+            << std::endl;
+}
+void ElectricCar::turnRight() const {
+  std::cout << __FUNCTION__ << ": " << __FILE__ << ": " << __LINE__
+            << std::endl;
+}
+void ElectricCar::brake() const {
+  std::cout << __FUNCTION__ << ": " << __FILE__ << ": " << __LINE__
+            << std::endl;
+}
+void ElectricCar::accelerate(Speed speed) const {
+  std::cout << __FUNCTION__ << ": " << __FILE__ << ": " << __LINE__
+            << std::endl;
+}
```

Comment:
you may move this behavior into ICar and use the common implementation. Code duplication is something that we want to avoid.